test/topology_experimental_raft/test_alternator.py

```diff
@@ -0,0 +1,203 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+# Multi-node tests for Alternator.
+#
+# Please note that most tests for Alternator are single-node tests and can
+# be found in the test/alternator directory. Most functional testing of the
+# many different syntax features that Alternator provides don't need more
+# than a single node to be tested, and should be able to run also on DynamoDB
+# - not just on Alternator, which the test/alternator framework allows to do.
+# So only the minority of tests that do need a bigger cluster should be here.
+
+import pytest
+import asyncio
+import logging
+import time
+import boto3
+import botocore
+import requests
+import json
+
+from test.pylib.manager_client import ManagerClient
+
+logger = logging.getLogger(__name__)
+
+# Convenience function to open a connection to Alternator usable by the
+# AWS SDK.
+alternator_config = {
+    'alternator_port': 8000,
+    'alternator_write_isolation': 'only_rmw_uses_lwt',
+    'alternator_ttl_period_in_seconds': '0.5',
+}
+def get_alternator(ip):
+    url = f"http://{ip}:{alternator_config['alternator_port']}"
+    return boto3.resource('dynamodb', endpoint_url=url,
+        region_name='us-east-1',
+        aws_access_key_id='alternator',
+        aws_secret_access_key='secret_pass',
+        config=botocore.client.Config(
+            retries={"max_attempts": 0},
+            read_timeout=300)
+    )
+
+# Alternator convenience function for fetching the entire result set of a
+# query into an array of items.
+def full_query(table, ConsistentRead=True, **kwargs):
+    response = table.query(ConsistentRead=ConsistentRead, **kwargs)
+    items = response['Items']
+    while 'LastEvaluatedKey' in response:
+        response = table.query(ExclusiveStartKey=response['LastEvaluatedKey'],
+            ConsistentRead=ConsistentRead, **kwargs)
+        items.extend(response['Items'])
+    return items
+
+# FIXME: boto3 is NOT async. So all tests that use it are not really async.
+# We could use the aioboto3 library to write a really asynchronous test, or
+# implement an async wrapper to the boto3 functions ourselves (e.g., run them
+# in a separate thread) ourselves.
+
+@pytest.fixture(scope="module")
+async def alternator3(manager_internal):
+    """A fixture with a 3-node Alternator cluster that can be shared between
+       multiple tests. These test should not modify the cluster's topology,
+       and should each use unique table names and/or unique keys to avoid
+       being confused by other tests.
+       Returns the manager object and 3 boto3 resource objects for making
+       DynamoDB API requests to each of the nodes in the Alternator cluster.
+    """
+    manager = manager_internal()
+    servers = await manager.servers_add(3, config=alternator_config)
```

Comment:
can't give proof from on top of my head, but it may be that each time you call  servers_add you disable cluster reuse feature, that's why configs are in suite.yaml

although I do think we need better support for cluster fixtures (like table fixtures in alternator/cql tests) in test.py long term, creating suite each time is not convenient

Reply:
short term, a question, why we can't enable alternator in the whole topology? or have a separate suite for alternator?

Reply:
Yes, the server reuse features in the topology* suites are terrible, so this is exactly what I wanted to solve here - not for the one test I'm writing, but for the future, to allow me to write 10 tests:
I added an "alternator3" fixture which will allow 1 or 10 or 50 tests to use this fixture and get a 3-node Alternator cluster which they all share, like in test/alternator. This is what we should have done all along - I really hate the idea of suite.yaml (all tests in the same directory use the same cluster), which in retrospect we can say did not work - almost non of the `topology*/*.py` tests reuse their cluster.
I want to change this, but it's hard to reuse a cluster when you just write one test ;-)

Reply:
> Alternator cluster which they all share, like in test/alternator. This is what we should have done all along - I really hate the idea of suite.yaml

Ok, I see. It's tricky because it introduces yet another caching mechanism.

But even worse it's incorrect form API point of view. Because then if you run tests concurrently and some will call add_server then the whole thing breaks.

Reply:
I don't understand what you mean. 
First, we do NOT run tests concurrently. Test.py runs things in parallel, but these are different invocations of pytest, etc., it's never several tests using the same Python objects in parallel.
Second, the whole idea of my "alternator3" fixture is that it would be used when you need a 3-node Alternator cluster which you DO NOT intend to change, so there is no risk that some tests will call add_server.

Eventually, when this framework has 1000 tests and not the 140 it has today, I predict that only a minority of these tests will want to do things like adding a node and removing a node. I predict that most of the tests will just ask for a 3-node (or whatever) cluster and use it, and 100 of these tests can use the same cluster. So we should make this case convenient and efficient.

Reply:
> so there is no risk that some tests will call add_server

there is a risk that someone without deep knowledge of test framework with add such server as API allows it but it's illegal and will fail in rather non trivial way.

> and 100 of these tests can use the same cluster

> it's never several tests using the same Python objects in parallel

aren't those two statements contradictory?

Reply:
Perhaps the solution for the alternator3 fixture to NOT give you back the manager object, just the Alternator objects.

But in general, my answer is no: Yes, when you share clusters and tables between tests, it's very easy to break things. Also in test/alternator, one test can for example change the configuration and break other tests. Or two tests can share a table and one of them enables a MV on it or CDC or changes the permissions or I don't know what. This is why there needs to be a policy on how good tests share, and not just code. I don't have the perfect policy-enforcing code, but I do no that not sharing anything - a la dtest - is NOT the solution. I absolutely love it that running the entire Alternator test suite (around 20,000 lines of test code) takes 30 seconds and not 30 minutes. I know this is unlikely to happen in test/topology, it's already "addicted" to slow tests, but I want to try turning the tide back a bit.